source/extensions/filters/http/grpc_shim/grpc_shim.cc

```diff
@@ -0,0 +1,170 @@
+    if (buffer.length() < Grpc::GRPC_FRAME_HEADER_SIZE) {
+      decoder_callbacks_->sendLocalReply(Http::Code::OK, "invalid request body", nullptr,
+                                         Grpc::Status::GrpcStatus::Unknown);
+      return Http::FilterDataStatus::StopIterationNoBuffer;
+    }
+
+    // Remove the gRPC frame header.
+    buffer.drain(Grpc::GRPC_FRAME_HEADER_SIZE);
+    prefix_stripped_ = true;
+  }
+
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterHeadersStatus GrpcShim::encodeHeaders(Http::HeaderMap& headers, bool) {
+  if (enabled_) {
+    auto content_type = headers.ContentType();
+
+    // If the response from upstream does not have the correct content-type,
+    // perform an early return with a useful error message in grpc-message.
+    if (content_type->value().getStringView() != upstream_content_type_) {
+      const auto grpc_message = fmt::format(
+          "envoy grpc-shim: upstream responded with unsupported content-type {}, status code {}",
+          content_type->value().getStringView(), headers.Status()->value().getStringView());
+
+      headers.insertGrpcMessage().value(grpc_message);
+      headers.insertGrpcStatus().value(Envoy::Grpc::Status::GrpcStatus::Unknown);
+      headers.insertStatus().value(enumToInt(Http::Code::OK));
+      content_type->value(content_type_);
+
+      return Http::FilterHeadersStatus::ContinueAndEndStream;
+    }
+
+    // Restore the content-type to match what the downstream sent.
+    content_type->value(content_type_);
+
+    // Adjust content-length to account for the frame header that's added.
+    adjustContentLength(headers, [](auto length) { return length + Grpc::GRPC_FRAME_HEADER_SIZE; });
+    // We can only insert trailers at the end of data, so keep track of this value
+    // until then.
+    grpc_status_ = grpcStatusFromHeaders(headers);
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus GrpcShim::encodeData(Buffer::Instance& buffer, bool end_stream) {
+  if (!enabled_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  if (end_stream) {
+    // Insert grpc-status trailers to communicate the error code.
+    auto& trailers = encoder_callbacks_->addEncodedTrailers();
+    trailers.insertGrpcStatus().value(grpc_status_);
+
+    // Compute the size of the payload and construct the length prefix.
+    //
+    // We do this even if the upstream failed: If the response returned non-200,
+    // we'll respond with a grpc-status with an error, so clients will know that the request
+    // was unsuccessful. Since we're guaranteed at this point to have a valid response
+    // (unless upstream lied in content-type) we attempt to return a well-formed gRPC
+    // response body.
+    const auto length = htonl(buffer.length() + buffer_.length());
+
+    std::array<uint8_t, Grpc::GRPC_FRAME_HEADER_SIZE> frame;
+    Grpc::Encoder().newFrame(Grpc::GRPC_FH_DEFAULT, htonl(length), frame);
```

Comment:
I don't think you need `htonl` here and above. `newFrame` takes care of that.

Reply:
Oh boy, I had just `length` here and it caused the tests to fail, didn't realize that I was already `htonl`ing on L146. At this point I'm just round tripping back to host 🤦‍♂️